ctc loss no batch dim support

george-qi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70092

Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

george-qi committed:
ctc loss no batch dim support

[ghstack-poisoned]

george-qi commented on test/test_nn.py:
```diff
@@ -6655,6 +6655,50 @@ def test_CTCLoss_zero_infinity(self):
+                                                             reduction=reduction, zero_infinity=True)
+
+                grad_gpu, = torch.autograd.grad(res_gpu.cuda(), log_probs.cuda(), grad_out.cuda())
+                grad_gpu_no_bd, = torch.autograd.grad(res_gpu_no_bd.cuda(), log_probs_no_bd.cuda(), grad_out.cuda())
```
I had some trouble running local tests for GPU on AWS -- it told me that there no GPU devices available. Do I need a devgpu to test this?

jbschlosser replied:
Try using `gpurun python test/test_nn.py -k <test_name>` with the definition from the AWS doc:
```
alias gpurun='srun -p q2 --cpus-per-task=16 -t 5:00:00 --gpus-per-node=2'
```

george-qi replied:
Tested with `gpurun` and it succeeded!

george-qi committed:
Update on "ctc loss no batch dim support"




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -36,13 +36,14 @@ static inline int64_t get_target_prime(target_t* target, int64_t offset, int64_t
 // The function returns the loss and the alphas, the alphas are kept for the backward step. The wrapper (ctc_loss below) hides
```
You need to undo the submodule checkin (android/libs/fbjni )

zou3519 commented on test/test_nn.py:
```diff
@@ -6655,6 +6655,50 @@ def test_CTCLoss_zero_infinity(self):
+            grad_out_no_bd = grad_out.detach().clone().double().requires_grad_()
+            grad_cpu_no_bd, = torch.autograd.grad(res_cpu_no_bd, log_probs_no_bd, grad_out_no_bd)
+
+            with torch.backends.cudnn.flags(enabled=False):
+                res_gpu = torch.nn.functional.ctc_loss(log_probs.cuda(), targets.cuda(), input_lengths, target_lengths,
+                                                       reduction=reduction, zero_infinity=True)
```
We have CPU-only builds of PyTorch. On those, .cuda() will error out with  "Torch not compiled with CUDA enabled".

There should be an`if USE_CUDA` (check me on the name of the variable) guard here

george-qi committed:
Update on "ctc loss no batch dim support"




[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"




[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"




[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280068).

jbschlosser commented on test/test_nn.py:
```diff
@@ -6830,6 +6830,57 @@ def test_CTCLoss_zero_infinity(self):
         self.assertEqual(g1, g2, atol=1e-4, rtol=0)
         self.assertTrue((g1 == g1).all().item())  # check that we don't have NaN
 
+    def test_CTCLoss_no_batch_dim(self):
```
I think it might be better to put this test under `TestNNDeviceType` instead of `TestNN`. This simplifies some things - you only have to write logic for a single device and you can use decorators from `torch/testing/_internal/common_device_type.py` such as `@disablecuDNN`.

Also, is it not possible to add a dict entries with a `single_batch_reference_fn` to `new_module_tests` like for the other modules?

george-qi replied:
Hmm, so the reason I didn't put it under `TestNNDeviceType` was because I wanted to make sure that the results were consistent between the different device types (that is, the logic was the same). If we did do device by device though, it would be much easier to use `single_batch_reference_fn` for each case though. Thoughts on this?

jbschlosser replied:
Ah I see - I think as long as `single_batch_reference_fn` passes for each device, there's no need to compare the results against each other. Note that you could alternatively create a `ModuleInfo` entry for `CTCLoss` and get the reference checks & CPU / GPU parity test for free. The `module_inputs_func` could return samples on `(device, device)` and an extra sample for the `(gpu, cpu)` case when the device is CUDA.

jbschlosser commented on aten/src/ATen/native/cudnn/LossCTC.cpp:
```diff
@@ -68,9 +68,10 @@ bool _use_cudnn_ctc_loss(
+std::tuple<Tensor, Tensor> _cudnn_ctc_loss(const Tensor& log_probs_t_, const Tensor& targets_t, IntArrayRef input_lengths_, IntArrayRef target_lengths_, int64_t BLANK, bool deterministic, bool zero_infinity) {
   (void)zero_infinity; // only used for backward
   const CheckedFrom c = "cudnn_ctc_loss";
+  Tensor log_probs_t = log_probs_t_.dim() == 3 ? log_probs_t_ : log_probs_t_.unsqueeze(1);
```
Can you explain a bit what's going with the cudnn version here? It looks like an unsqueeze is being done but there is no matching resqueeze for the output - am I missing something?

george-qi replied:
Thanks for the catch! Added a test that catches this and fixed the bug

jbschlosser reviewed (commented):
Thanks for the update!

I didn't see any doc updates, but I think we need those as well

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented on test/test_nn.py:
```diff
@@ -6830,6 +6830,70 @@ def test_CTCLoss_zero_infinity(self):
+                    input_lengths_no_bd.cuda(),
+                    target_lengths_no_bd.cuda(),
+                    zero_infinity=True
+                )                    
```
This now successfully accesses `_cudnn_ctc_loss`

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -36,13 +36,14 @@ static inline int64_t get_target_prime(target_t* target, int64_t offset, int64_t
   using target_t = typename std::conditional<target_scalar_type == kInt, int, int64_t>::type;
 
   CheckedFrom c = "ctc_loss_cpu";
+  Tensor log_probs = log_probs_.dim() == 3 ? log_probs_ : log_probs_.unsqueeze(1);
```
I keep getting confused by these unsqueezes without corresponding squeezes :p

Proposal: handle batching at the highest level [`ctc_loss` function](https://github.com/pytorch/pytorch/blob/b15212c62bb00b1153f35e78b8eaa022f9dd867f/aten/src/ATen/native/LossCTC.cpp#L369-L405). There should be no need to touch the CPU / GPU specific impls if you do this. Wdyt?

george-qi replied:
Also works! Probably a bit cleaner, let me do that :)

jbschlosser reviewed (commented):
Added a suggestion that may help simplify the logic. TL;DR: maybe squeeze / unsqueeze logic can be done in `ctc_loss` only?

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280068).

jbschlosser commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -171,10 +171,12 @@ std::tuple<Tensor, Tensor> ctc_loss_cpu_template(const Tensor& log_probs, const
                                       const Tensor& neg_log_likelihood, const Tensor& log_alpha, int64_t BLANK, bool zero_infinity) {
   constexpr scalar_t neginf = -std::numeric_limits<scalar_t>::infinity();
   using target_t = typename std::conditional<target_scalar_type == kInt, int, int64_t>::type;
+  auto is_batched = log_probs_.dim() == 3;
+  Tensor log_probs = is_batched ? log_probs_ : log_probs_.unsqueeze(1);
```
This looks pretty much like I'd expect for the backward pass - unsqueeze -> do stuff -> squeeze at the end. There's one thing I'm not sure about though: since `grad_out` is an output tensor, are we sure it's sized correctly at this point?

jbschlosser commented on aten/src/ATen/native/cuda/LossCTC.cu:
```diff
@@ -739,6 +740,9 @@ Tensor ctc_loss_backward_gpu_template(const Tensor& grad_out, const Tensor& log_
 
+  if (!is_batched) {
+    return grad.squeeze(1);
+  }
```
Honestly I'm still a bit confused about the changes to this file - if we handle things above this at the `ctc_loss` level, I wouldn't expect to need changes here. Can you help me understand this a bit more?

jbschlosser commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -369,7 +375,8 @@ Tensor ctc_loss_backward_cpu(const Tensor& grad, const Tensor& log_probs, const
 // it also handles the reduction if desired
-Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t BLANK, int64_t reduction, bool zero_infinity) {
+Tensor ctc_loss(const Tensor& log_probs_, const Tensor& targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t BLANK, int64_t reduction, bool zero_infinity) {
+  Tensor log_probs = log_probs_.dim() == 3 ? log_probs_ : log_probs_.unsqueeze(1);
```
Where does the corresponding squeeze for this unsqueeze happen?

jbschlosser commented on test/test_nn.py:
```diff
@@ -16404,6 +16404,81 @@ def _helper(zero_infinity):
         _helper(zero_infinity=True)
         _helper(zero_infinity=False)
 
+    def _CTCLoss_no_batch_dim_helper(self, device):
```
I think longer term the better approach would be to write a ModuleInfo entry for CTCLoss covering these cases, but I'm okay with these tests for now.

jbschlosser commented on test/test_nn.py:
```diff
@@ -16404,6 +16404,81 @@ def _helper(zero_infinity):
+        grad_no_bd, = torch.autograd.grad(res_no_bd, log_probs_no_bd, grad_out_no_bd)
+
+        with torch.backends.cudnn.flags(enabled=True):
+            res_mixed = torch.nn.functional.ctc_loss(log_probs, targets.cpu(), input_lengths, target_lengths, zero_infinity=True)
+            res_no_bd_mixed = torch.nn.functional.ctc_loss(
```
I think it's slightly better to test the module form of `CTCLoss` in this test and below as this will also test the functional form.

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented on aten/src/ATen/native/cudnn/LossCTC.cpp:
```diff
@@ -135,6 +135,7 @@ std::tuple<Tensor, Tensor> _cudnn_ctc_loss(const Tensor& log_probs_t, const Tens
       ctc_loss_desc.desc(),
       workspace.data_ptr(),
       workspace_size));
+
```
Did this to keep it consistent with the spacing in the other files, but I'll remove it!

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented on test/test_nn.py:
```diff
@@ -16404,6 +16404,123 @@ def _helper(zero_infinity):
+                losses.append(func_ctc(log_probs_refs[0], targets.cuda(), input_lengths, target_lengths, reduction=reduction, zero_infinity=True))
+                losses.append(func_ctc(log_probs_refs[1], targets_no_bd.cuda(), input_lengths, target_lengths, reduction=reduction, zero_infinity=True))
+                losses_no_bd.append(func_ctc(log_probs_no_bd_refs[0], targets_no_bd.cuda(), input_lengths_no_bd,
+                                             target_lengths_no_bd, reduction=reduction, zero_infinity=True))
```
@jbschlosser Updated the testing to check the following cases

1. `log_probs.shape = (T, N, C)`, `targets.shape = (N, S)`
2. `log_probs.shape = (T, N, C)`, `targets.shape = (S)`
3. `log_probs.shape = (T, C)`, `targets.shape = (S)`

george-qi commented on test/test_nn.py:
```diff
@@ -16404,6 +16404,123 @@ def _helper(zero_infinity):
+        _check_list(losses[0].squeeze(0), losses_no_bd, atol=1e-4, rtol=0)
+
+        _check_list(log_probs_refs[0].grad, [t.grad for t in log_probs_refs[1:]], atol=1e-4, rtol=0)
+        _check_list(log_probs_refs[0].grad.squeeze(1), [t.grad for t in log_probs_no_bd_refs[1:]], atol=1e-4, rtol=0)
```
@jbschlosser This function checks each combination of the size/values of the losses/gradients against each other

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

jbschlosser commented on test/test_nn.py:
```diff
@@ -16431,6 +16431,123 @@ def _helper(zero_infinity):
         _helper(zero_infinity=True)
         _helper(zero_infinity=False)
 
+    def _CTCLoss_no_batch_dim_helper(self, device, input_length, vocab_size, target_length):
```
The test logic is a bit tricky to validate since there's a lot going on. Rather than iterate through the various cases within a single test, I think it could be simplified quite a bit by doing a couple things:
* Write test logic for a single case (create unbatched & batched inputs, run both through, compare) and using `@parametrize_test` to iterate over the cases instead. I can definitely help with this if needed!
* No need to explicitly duplicate cases for the cudnn case. We usually write a CUDA test and assume that CI will run it in both cuDNN and non-cuDNN environments. If there is cudnn-specific stuff that needs to be done (e.g. place targets on CPU), you can check if cuDNN is enabled and do it then.
* You can also use `@parametrize_test("reduction", ["mean", "sum", "none"])` so you don't need to iterate over reductions in the test.

george-qi replied:
- Incorporated `@parametrize_test` 
- Consolidated the CPU and CUDA tests by refactoring a bit and then split the shape checking / value checking tests
I think these together make it indeed look much cleaner!

jbschlosser reviewed (commented):
Some comments on simplifying the test logic below, doc & functional changes lgtm :)

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented on torch/nn/modules/loss.py:
```diff
@@ -1648,7 +1648,7 @@ class CTCLoss(_Loss):
           targets, the target_lengths must add up to the total length of the tensor.
         - Output: scalar. If :attr:`reduction` is ``'none'``, then
-          :math:`(N)`, where :math:`N = \text{batch size}`.
+          :math:`(N)` or :math:`()`, where :math:`N = \text{batch size}`.
```
Done, thanks :)

george-qi commented on test/test_nn.py:
```diff
@@ -16431,6 +16431,123 @@ def _helper(zero_infinity):
+            log_probs_no_bd_refs = [log_probs_no_bd.detach().clone().requires_grad_() for _ in range(2)]
+
+            func_ctc = torch.nn.functional.ctc_loss
+            ctc_loss = nn.CTCLoss(reduction=reduction, zero_infinity=True)
```
Added a few comments!

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi commented on torch/nn/modules/loss.py:
```diff
@@ -1648,7 +1648,7 @@ class CTCLoss(_Loss):
-          :math:`(N)`, where :math:`N = \text{batch size}`.
+          :math:`(N)` or :math:`()`, where :math:`N = \text{batch size}`.
 
     Examples::
```
Good idea - added an example!

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

jbschlosser commented on test/test_nn.py:
```diff
@@ -16456,6 +16456,118 @@ def _helper(zero_infinity):
+        target_length = 12
+
+        # cuda/cudnn case
+        if torch.cuda.is_available():
```
To simplify the device placement logic, I think you can just do this:
1. Use `device=device` for both input and target
2. If torch.cuda.is_available() and torch.backends.cudnn.enabled, move target to CPU (i.e. `target = target.cpu()`)
3. Do function / module calls with input + target
4. Do checks

As another possibly simplifying thing, you could also do:
```
@parametrize_test("use_module_form", [True, False])
```

and call the functional form if `use_module_form=False`

jbschlosser reviewed (commented):
Pretty close! Couple minor suggestions then we can get this wrapped up :)

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

george-qi committed:
Update on "ctc loss no batch dim support"


Differential Revision: [D33280068](https://our.internmc.facebook.com/intern/diff/D33280068)

[ghstack-poisoned]

jbschlosser commented on test/test_nn.py:
```diff
@@ -16456,6 +16456,91 @@ def _helper(zero_infinity):
+
+        return losses, losses_no_bd, log_probs_refs, log_probs_no_bd_refs
+
+    def _assertEqual_list(self, expected, list_to_compare, atol=None, rtol=None):
+        for ele in list_to_compare:
+            self.assertEqual(expected, ele, atol=atol, rtol=rtol)
```
I think `self.assertEqual()` handles lists already, no?

jbschlosser reviewed (approved):
LGTM! Thanks for the update :)

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280068).
